rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/telemetry/TbMsgDeleteTimeseriesNode.java

```diff
@@ -0,0 +1,115 @@
+import org.thingsboard.server.common.data.kv.BaseDeleteTsKvQuery;
+import org.thingsboard.server.common.data.kv.DeleteTsKvQuery;
+import org.thingsboard.server.common.data.plugin.ComponentType;
+import org.thingsboard.server.common.msg.TbMsg;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+@RuleNode(
+        type = ComponentType.ACTION,
+        name = "delete timeseries",
+        configClazz = TbMsgDeleteTimeseriesNodeConfiguration.class,
+        nodeDescription = "Delete timeseries data for Message Originator.",
+        nodeDetails = "In case that no telemetry keys are selected - message send via <b>Failure</b> chain. " +
+                "In order to delete all data for keys checkbox <b>Delete all data for keys<b> should be selected, " +
+                "otherwise will be deleted data that is in range of the selected time interval. " +
+                "In order to rewrite the latest values for keys selected to delete, checkbox <b>Rewrite latest if deleted<b> " +
+                "should be selected! If data for selected timeseries keys successfully deleted - " +
+                "message send via <b>Success</b> chain, otherwise <b>Failure</b> chain will be used.",
+        uiResources = {"static/rulenode/rulenode-core-config.js"},
+        configDirective = "tbActionNodeDeleteTimeseriesConfig",
+        icon = "remove_circle"
+)
+public class TbMsgDeleteTimeseriesNode implements TbNode {
+
+    private TbMsgDeleteTimeseriesNodeConfiguration config;
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        this.config = TbNodeUtils.convert(configuration, TbMsgDeleteTimeseriesNodeConfiguration.class);
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) throws ExecutionException, InterruptedException, TbNodeException {
+        List<String> keysPatterns = config.getKeysPatterns();
+        if (CollectionUtils.isEmpty(keysPatterns)) {
+            ctx.tellFailure(msg, new IllegalStateException("Telemetry keys list is not selected!"));
+        } else {
+            long deleteFromTs;
+            long deleteToTs;
+            long currentTimeMillis = System.currentTimeMillis();
+            try {
+                if (config.getDeleteAllDataForKeys()) {
+                    deleteFromTs = 0L;
+                    deleteToTs = currentTimeMillis;
+                } else {
+                    if (config.isUseMetadataIntervalPatterns()) {
+                        deleteFromTs = setDeleteTsIntervalFromPattern(msg, config.getStartTsIntervalPattern());
+                        deleteToTs = setDeleteTsIntervalFromPattern(msg, config.getEndTsIntervalPattern());
+                    } else {
+                        deleteFromTs = currentTimeMillis - TimeUnit.valueOf(config.getStartTsTimeUnit()).toMillis(config.getStartTs());
```

Comment:
Calculate 'TimeUnit.valueOf(config.getStartTsTimeUnit()).toMillis(config.getStartTs()' once in the init method. Not for each message